pkg/store/proxy.go

```diff
@@ -344,11 +411,49 @@ func startStreamSeriesSet(
 
 	wg.Add(1)
 	go func() {
+		metricsSended := false
 		defer wg.Done()
 		defer close(s.recvCh)
-
 		for {
-			r, err := s.stream.Recv()
+			cctx := ctx
+			var cancel context.CancelFunc
+			timeoutMsg := fmt.Sprintf("failed to receive any data from %s", s.name)
```

Comment:
I don't like the idea to prepare error message in the very beginning. The only difference is in line #422: responseTimeout is added. But do we really need it? As I know it's configuration value. So it can be built on #442.